### PR TITLE
Add Scalar methods and operators

### PR DESCRIPTION
**Pull request by conradoplg**
This is on top of #22 so it's including the fixes there.

This adds code required to use this crate with https://github.com/ZcashFoundation/frost/ 

- `Scalar::from_canonical_bytes`
- `Scalar::bytes_mod_order_wide` which is required by
- `Scalar::random`
- Operators, with code lifted from curve25519-dalek.

### CODE REVIEW

File: src/field/scalar.rs [hunk 1]
```diff
@@ -260,6 +265,71 @@ impl Scalar {
...
+    ///
+    /// - `Some(s)`, where `s` is the `Scalar` corresponding to `bytes`,
+    ///   if `bytes` is a canonical byte representation;
+    /// - `None` if `bytes` is not a canonical byte representation.
+    pub fn from_canonical_bytes(bytes: [u8; 57]) -> Option<Scalar> {
+        // Check that the 10 high bits are not set
+        if bytes[56] != 0u8 || (bytes[55] >> 6) != 0u8 {
+            return None;
+        }
+        let bytes: [u8; 56] = std::array::from_fn(|i| bytes[i]);
```
Comment (kevaundray) [inline src/field/scalar.rs:281]: First time seeing `std::array::from_fn`, learn something new each day :)

**Review by kevaundray — APPROVED** [review]

**Comment by conradoplg** [comment]
Rebased after the previous PR got merged (thanks!)

File: src/field/scalar.rs (hunk 1 above)
Comment (conradoplg) [reply src/field/scalar.rs:281]: I just learned about it myself, while searching for a less annoying way to get a subarray from a array :grin: 

BTW --- RFC 8032 specifies that scalars should be serialized as 57 bytes, to match the point serialization length. Is this a change you're interested in?

Comment (conradoplg) [reply src/field/scalar.rs:281]: (I mean, in `to/from_bytes` and not just `from_canonical_bytes`)

Comment (kevaundray) [reply src/field/scalar.rs:281]: Yes very interested!

**Comment by kevaundray** [comment]
Will merge, feel free to open a PR for further edits, thanks again :)

**Pull request merged**